택배송장 [STEP 1] tony lim

witlim1225 opened the pull request:
@havilog 
     드디어 첫 STEP의 PR을 보내네요 🙌
     질문 양식은... yagom의 도움을 받도록 할게요. 
     코드의 주석도 여러가지 써봤는데 도움이 됐으면 좋겠어요.

      ## 고민했던 점
      ### 주제
      - 이번 과제에서는 우선 yagom의 힌트를 적극적으로 순서를 정하고 
      - 왜 해당 요구사항을 엔터를 쳐서 분리시켰을까 하는 생각을 하면서
      - 코드에 접근하려고 했어요.
      - HINT에서 했던 방식을 토대로 추가 요구사항에도 반영해보려고 노력했습니다.

      ---

      ## 해결이 되지 않은 점
      ### 주제
      - 처음엔 잘 이해가 되지 않았지만, protocol을 통해 실제로 해당 프로퍼티를 소유하지 않고
      - init시 생성 혹은 주입을 받을 수 있도록 하는 방식으로 변경함으로써
      - 코드에서의 결합(?) 소유(?)가 약해졌다는 느낌을 받았어요.
      - 초기라 그런지 시키는 대로 STEP 1을 진행하면서 이런게 DIP인가?! 하는,
      - 코드로는 이해가 가지만, 이걸 다음에 다시 적용할 수 있는가에 대한 의문이 남아요.
      - 이런 건 다음 STEP 을 진행하다보면 나아질 지 궁금함이 있어요.
      - (여기서도 조언을 얻고 싶은 부분이 나와버리네요)
     
      ---

      ## 조언을 얻고 싶은 부분
      ### 주제
      - 막연하게 SRP는 `한가지 기능만 해야지!`라고 생각하고 
      - 일단 코드를 분리해 버리기도 하는데, 이게 제대로 하는건지 잘 모르겠어요.
      - 더불어서 SOLID 원칙을 생각하면서 코드를 짤 떄, 
      - `이건 어떤 원칙을 지키기 위해서 이렇게  코드를 구성하고 있는거야`라는 생각이 드는지 궁금해요.
      - 마지막으로, 기획자가 의도한 대로 동작은 하게 만들지만, 
      - 처음에 내가 생각한 대로의 디자인패턴인지, 어떤 원칙을 지켜서 하려고 했는데
      - 실제로도 코드가 그렇게 작성한건지. 이런 부분들 확인은 어떻게 확인할 수 있는지 궁금해요. 
      - ( iOS가 1명이라 이런 부분들에 있어서 불안함이 있어요 )

witlim1225 committed:
택배송장 [STEP 1] tony lim

havilog commented on ParcelInvoiceMaker/ParcelInvoiceMaker/InvoiceView.swift:
```diff
@@ -24,17 +24,20 @@ class InvoiceView: UIView {
         nameLabel.textColor = .black
         nameLabel.font = .preferredFont(forTextStyle: .largeTitle)
-        nameLabel.text = "이름 : \(parcelInformation.receiverName)"
+        //nameLabel.text = "이름 : \(parcelInformation.receiverName)"
```
기존 코드를 특별히 남겨두신 이유가 있으실까요??

witlim1225 replied:
우선 초반이라 내가 어떤식으로 바꿨는지 체크하기 위해서 남겨둔 부분인데, 다음 STEP에서는 지워서 피알 올릴게요.
지금 보니 굳이 남길 필요까진 없을거 같아요.

havilog commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelOrderView.swift:
```diff
@@ -76,11 +76,20 @@ class ParcelOrderView: UIView {
                                                          discount: discount)
+         */
+        
+        let userInfo: UserInfo = .init(address: address, receiverName: name, receiverMobile: mobile)
+        
+        let parcelInformation: ParcelInformation = .init(userInfo: userInfo,
+                                                         deliveryCost: cost,
+                                                         discount: discount)
```
파라미터 줄바꿈을 하시는 기준이 있으실까요?
저 같은 경우에는 3개 이상인 경우에 하는 편이에요!

witlim1225 replied:
아하~ 저는 한 화면에 안보일 정도가 되면 줄바꿈 하는 타입이었던거 같아요.
가끔 애매하게 길어져서 고민할 때 있었는데 '3개'같은 기준을 들으니 좋아 보이네요.
(parcelInfomation)같은 경우는 기존의 5개 코드가 있었던 방식을 그대로 놔둔거라 저렇게 놔두게 된거였는데요.
이렇게 보니 확실히 통일 시켜주면 좋았을거 같네요.

havilog commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelOrderViewController.swift:
```diff
@@ -7,9 +7,13 @@
+    private let parcelProcessor: ParcelOrderProcessorPersistence
     
     init() {
+        // 인스턴스를 여기서는 생성했지만, 주입받을 수도 있도록 변경
+        self.parcelProcessor = ParcelOrderProcessor()
```
> 주입받을 수도 있도록 변경

은 앞으로 하실 예정이라 작성해놓으신 걸까요?

만약 아니라면,
Dependency Injection을 위해서 init시점에 파라미터를 protocol로 받아, 외부에서 의존성을 주입해보는건 어떨까요?

witlim1225 replied:
현재 STEP에서 아직 해당 단계까지 가지 않은 거같아서, 일부러 TODO처럼 주석을 남겨놨습니다.
해당 사항에 대해서는 STEP 2 PR에 반영해볼게요~

havilog commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelOrderViewController.swift:
```diff
@@ -18,13 +22,22 @@ class ParcelOrderViewController: UIViewController, ParcelOrderViewDelegate {
     
+    // 택배 주문을 만드는 곳인데 1)프로세스를 진행하고나서 2) 송장번호 확인하는 페이지까지 이동시키네 -> 분리해보자
     func parcelOrderMade(_ parcelInformation: ParcelInformation) {
         parcelProcessor.process(parcelInformation: parcelInformation) { (parcelInformation) in
```
괄호 ()가 필요없다면 지워보는건 어떨까요?

witlim1225 replied:
그냥 지나쳤던 부분인데 알려주셔서 감사합니다~

witlim1225 replied:
괄호를 지우면서 직접 함수를 전달해주는게 더 좋지 않을까 생각해서, 수정 해봤습니다.

havilog replied:
함수를 전달해주는게 좋은 경우도 있는 것은 공감해요!
다만 코드 점핑이 한 번 더 일어난다거나, self를 약하게 참조해야하는 경우에 빼먹을 수도 있어서
TPO에 맞게 전달하시면 좋을듯 합니다 ㅎㅎ

havilog commented on ParcelInvoiceMaker/ParcelInvoiceMaker/ParcelProcessor.swift:
```diff
@@ -23,30 +37,108 @@ class ParcelInformation {
         onComplete(parcelInformation)
     }
     
+    /*
+     * 1. 데이터 베이스에 저장하는 기능 -> DatabaseParcelInformationPersistence로 이동
+    func save(parcelInformation: ParcelInformation) {
+        // 데이터베이스에 주문 정보 저장
+        print("발송 정보를 데이터베이스에 저장했습니다.")
+    }
+     */
+}
+
+// 3. 1). i) ParcelInformationPersistence 프로토콜 지정
+protocol ParcelInformationPersistence {
+    // 3. 2) ii)를 진행하고나서 추가한 함수.
+    func save(parcelInformation: ParcelInformation)
+}
```
주석, 기존에 있던 함수들이 코드의 가독성을 떨어트리는 느낌이 들어요!
반드시 필요한 주석만 남겨보는건 어떨까요?

witlim1225 replied:
저부분의 경우 학습용으로 남겨둔 코드였어요.
미처 저걸 생각하지 못한 반성의 의미로...ㅎㅎ
다음 STEP에서는 기록에 남았으니 지우고 PR을 보내도록 하겠습니다!

havilog reviewed (approved):
> 해결이 되지 않은 점

구체타입에 의존하지 않고, protocol에 의존하게 되면,
인터페이스에만 의존하기 때문에 실제 구현부가 바뀌더라도, 인터페이스를 사용하는 쪽의 코드변화가 적거나 없기 때문에 결합도가 낮아진다고 볼 수 있을거 같아요.
또한 DIP원칙을 지킴으로써, Dependency Injection을 할 수 있다는게 장점이라고 느껴지는데요,
요 부분은 검색해보시면 좋을 듯 합니다!

> SRP에 대한 고민

지금 스탭에서는 사실 요 원칙을 어길만큼 많은 동작을 하고 있는 객체는 없는거 같아요.
어떠한 객체가 할 일의 범위를 잘 정하는 것은 개발자의 역량이고 정답은 없는 것 같습니다!
다음 스탭들을 진행하면서 경험해나가면 좋을 듯 싶어요

> 실제로도 코드가 그렇게 작성한건지. 이런 부분들 확인은 어떻게 확인할 수 있는지 궁금해요.

요건 사실 질문의 의도를 
"혼자 개발하고 있는데, 잘 하고 있는 것인지 불안하다"
라고 이해했는데 제가 이해한 부분이 맞을까요?

맞다면, 이 부분은 3주간 코드를 같이 보면서 어떤 코드가 지향해야하고, 어떤 코드는 안티패턴인지 등을 논의해보면 될 듯 한데요,
어떤 디자인패턴과 아키텍쳐가 정답이다!는 없는거 같고,
지금 시점보다는 캠프가 끝날 때 한 번 더 얘기해보는 것도 좋을 것 같습니다 ㅎㅎ

witlim1225 committed:
택배송장 [STEP 1] tonylim - feedback

- 불필요한 주석 삭제
- 파라미터가 3개 이상인 경우 / 한 화면에서 잘리는 긴 파라미터 속성의 경우 줄바꿈 처리
- init시점에 파라미터를 protocol로 받아와 외부에서 의존성을 주입할 수 있도록 처리
- 불필요한 괄호() 를 삭제 -> 이를 진행하며, 함수를 직접 전달하는 부분으로 수정 처리

witlim1225 commented:
@havilog
해당 PR이 닫힌줄 착각하고 있었네요.
코멘트 달아주신 부분에 대해서 다시 수정해서 올렸습니다.
